### ISSUE TITLE
Introduce static analysis

### ISSUE THREAD

**Issue by szepeviktor** [johnbillion/query-monitor#508]
`phpstan.neon.dist`

```yaml
includes:
    - vendor/szepeviktor/phpstan-wordpress/extension.neon
parameters:
    level: max
    inferPrivatePropertyTypeFromConstructor: true
    autoload_files:
        - tests/phpstan/bootstrap.php
        - classes/QM.php
        - classes/Plugin.php
        # All other classes
        - classes/Activation.php
        - classes/Backtrace.php
        - classes/Collector.php
        - classes/Collectors.php
        - classes/debug_bar_panel.php
        - classes/debug_bar.php
        - classes/Dispatcher.php
        - classes/Dispatchers.php
        - classes/Hook.php
        - classes/Output.php
        - classes/QueryMonitor.php
        - classes/Timer.php
        - classes/Util.php
        - collectors/php_errors.php
    autoload_directories:
        - classes/
        - collectors/
        - dispatchers/
        - output/
    paths:
        - classes/
        - collectors/
        - dispatchers/
        - output/
    ignoreErrors:
        # Uses func_get_args()
        - '#^Function apply_filters(_ref_array)? invoked with [34567] parameters, 2 required\.$#'
        # TODO Define properties
        - '#^Access to an undefined property QM_#'
```

`tests/phpstan/bootstrap.php`

```php
Oh! I've deletes the query-monitor directory without copying this file here!
It contained
- core constants not listed in phpstan-wordpress
- some 3rd-party functions stubs
- QM constants
```

```
composer require --dev szepeviktor/phpstan-wordpress
vendor/bin/phpstan analyze -l 2
```

Then raise level!

**Comment by szepeviktor** [comment]
Current list of problems on Level 2.

```
 ------ --------------------------------------------------
  Line   classes/QM.php
 ------ --------------------------------------------------
  108    Call to an undefined method QM_Collector::log().
 ------ --------------------------------------------------

 ------ ------------------------------------------------------------------------
  Line   classes/Util.php
 ------ ------------------------------------------------------------------------
  109    Variable $type might not be defined.
  111    Variable $type might not be defined.
  123    Variable $type might not be defined.
  135    Variable $type might not be defined.
  143    Variable $type might not be defined.
  182    Call to function compact() contains possibly undefined variable $type.
 ------ ------------------------------------------------------------------------

 ------ ----------------------------------------------------------------
  Line   collectors/caps.php
 ------ ----------------------------------------------------------------
  78     PHPDoc tag @param references unknown parameter: $required_caps
 ------ ----------------------------------------------------------------

 ------ ---------------------------------------------------------------------------
  Line   collectors/environment.php
 ------ ---------------------------------------------------------------------------
  118    Function mysqli_get_client_version invoked with 0 parameters, 1 required.
 ------ ---------------------------------------------------------------------------

 ------ ------------------------------------------------
  Line   collectors/request.php
 ------ ------------------------------------------------
  242    Access to an undefined property object::$slug.
  252    Access to an undefined property object::$name.
 ------ ------------------------------------------------

 ------ ---------------------------
  Line   collectors/theme.php
 ------ ---------------------------
  128    Constant WPINC not found.
 ------ ---------------------------

 ------ ----------------------------------------------------------------------------------------
  Line   dispatchers/Html.php
 ------ ----------------------------------------------------------------------------------------
  15     Property QM_Dispatcher_Html::$outputters has unknown class QM_Output_html as its type.
  231    Call to method output() on an unknown class QM_Output_html.
  237    Call to method set_timer() on an unknown class QM_Output_html.
  281    Call to an undefined method (object)|(object)::id().
  535    Variable $current_value might not be defined.
 ------ ----------------------------------------------------------------------------------------

 ------ -----------------------------------
  Line   dispatchers/WP_Die.php
 ------ -----------------------------------
  49     Variable $i might not be defined.
  50     Variable $i might not be defined.
  52     Variable $i might not be defined.
 ------ -----------------------------------

 ------ ---------------------------------------------------
  Line   output/Html.php
 ------ ---------------------------------------------------
  22     Call to an undefined method QM_Collector::name().
 ------ ---------------------------------------------------

 ------ ---------------------------------------------------------
  Line   output/html/debug_bar.php
 ------ ---------------------------------------------------------
  83     Call to an undefined method QM_Collector::is_visible().
 ------ ---------------------------------------------------------
```

**Comment by szepeviktor** [comment]
Some places you have to manually tell the type of result of `QM_Collectors::get()`

```php
/** @var QM_Collector_Logger $logger */
$logger = QM_Collectors::get( 'logger' );
```

**Comment by johnbillion** [comment]
Fixed in #625
